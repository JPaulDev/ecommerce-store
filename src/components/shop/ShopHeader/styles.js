import styled, { css } from 'styled-components';

const transition = css`
  transform: ${({ isOpen }) => (isOpen ? 'rotate(-90deg)' : 'rotate(90deg)')};
  transition-delay: 500ms;
  transition-property: transform;
`;

const containerStyles = css`
  align-items: center;
  display: flex;
  justify-content: space-between;
`;

const buttonStyles = css`
  cursor: pointer;
  font-size: 0;
`;

export const TopContainer = styled.div`
  border-bottom: 3px solid var(--color-grey-12);
  padding-bottom: 20px;
  ${containerStyles}
`;

export const Heading = styled.h1`
  font-family: 'Uni Sans', sans-serif;
  font-size: var(--font-size-30);
  margin-right: 35px;
`;

export const TopButton = styled.button`
  display: flex;
  ${transition}
  ${buttonStyles}

  @media screen and (min-width: 650px) {
    display: none;
  }
`;

export const BottomContainer = styled.div`
  margin-top: 15px;

  @media screen and (max-width: 649px) {
    ${containerStyles}
  }
`;

export const Subheading = styled(Heading)`
  display: inline-block;
  font-size: var(--font-size-22);
  font-weight: var(--font-weight-semibold);

  @media screen and (min-width: 700px) {
    margin-right: 20px;
  }
`;

export const BottomButton = styled.button`
  align-items: center;
  display: inline-flex;
  font-weight: var(--font-weight-bold);
  text-transform: uppercase;
  ${buttonStyles}

  svg {
    height: auto;
    width: 15px;
    ${transition}
  }

  @media screen and (min-width: 650px) {
    column-gap: 7px;
    color: var(--color-blue-4);
    font-size: var(--font-size-13);

    svg {
      height: auto;
      width: 10px;
    }
  }
`;

export const Text = styled.p`
  color: var(--color-grey-4);
  font-size: var(--font-size-14);
  line-height: 1.7;
  padding-top: 15px;
`;
