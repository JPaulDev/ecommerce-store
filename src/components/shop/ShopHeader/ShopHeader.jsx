import { motion, useCycle } from 'framer-motion';
import Responsive from '../../common/Responsive';
import { ChevronRight } from '../../icons';
import * as Styled from './styles';

const motionProps = {
  style: { overflow: 'hidden' },
  initial: false,
  variants: {
    open: {
      height: 'auto',
      display: 'block',
    },
    closed: {
      height: 0,
      transitionEnd: {
        display: 'none',
      },
    },
  },
  transition: {
    ease: 'easeInOut',
    duration: 0.6,
  },
};

export default function ShopHeader() {
  const [isTopOpen, toggleTopOpen] = useCycle(false, true);
  const [isBottomOpen, toggleBottomOpen] = useCycle(false, true);

  return (
    <>
      <Styled.TopContainer>
        <Styled.Heading>Placeholder</Styled.Heading>
        <Styled.TopButton
          type="button"
          isOpen={isTopOpen}
          aria-controls="top-text"
          aria-expanded={isTopOpen}
          onClick={toggleTopOpen}
        >
          Read More
          <ChevronRight width={20} height={27} />
        </Styled.TopButton>
      </Styled.TopContainer>
      <Responsive sx={{ display: { 0: 'none', 650: 'block' } }}>
        <Styled.Text>Placeholder</Styled.Text>
      </Responsive>
      <Responsive sx={{ display: { 0: 'block', 650: 'none' } }}>
        <motion.div
          id="top-text"
          animate={isTopOpen ? 'open' : 'closed'}
          {...motionProps}
        >
          <Styled.Text>Placeholder</Styled.Text>
        </motion.div>
      </Responsive>
      <Styled.BottomContainer>
        <Styled.SubHeading as="h2">Placeholder</Styled.SubHeading>
        <Styled.BottomButton
          type="button"
          isOpen={isBottomOpen}
          aria-controls="bottom-text"
          aria-expanded={isBottomOpen}
          onClick={toggleBottomOpen}
        >
          Read More
          <ChevronRight />
        </Styled.BottomButton>
      </Styled.BottomContainer>
      <motion.div
        id="bottom-text"
        animate={isBottomOpen ? 'open' : 'closed'}
        {...motionProps}
      >
        <Styled.Text>Placeholder</Styled.Text>
      </motion.div>
    </>
  );
}
