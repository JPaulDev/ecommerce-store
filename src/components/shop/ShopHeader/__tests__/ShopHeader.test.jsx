import {
  render,
  screen,
  waitFor,
  waitForElementToBeRemoved,
} from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import renderer from 'react-test-renderer';
import ShopHeader from '../index';

const categoryData = {
  heading: 'heading',
  description: 'category description',
  latestArrivals: 'latest arrivals',
  latestArrivalsDescription: 'latest arrivals description',
};

describe('ShopHeader component', () => {
  it('should render a category heading and description', () => {
    const tree = renderer.create(<ShopHeader />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('ShopHeader component description and latest arrivals', () => {
  it('should open and close the category description', async () => {
    render(<ShopHeader {...categoryData} />);
    const text = /category description/i;
    const button = screen.getByTestId('description-button');

    expect(screen.getAllByText(text)).toHaveLength(1);

    userEvent.click(button);

    expect(screen.getAllByText(text)).toHaveLength(2);

    userEvent.click(button);

    await waitFor(() => {
      expect(screen.getAllByText(text)).toHaveLength(1);
    });
  });

  it('should open and close the latest arrivals description', async () => {
    render(<ShopHeader {...categoryData} />);
    const text = /latest arrivals description/i;
    const button = screen.getByTestId('latest-arrivals-button');

    expect(screen.queryByText(text)).not.toBeInTheDocument();

    userEvent.click(button);

    expect(screen.getByText(text)).toBeInTheDocument();

    userEvent.click(button);

    await waitForElementToBeRemoved(screen.queryByText(text));
  });
});
