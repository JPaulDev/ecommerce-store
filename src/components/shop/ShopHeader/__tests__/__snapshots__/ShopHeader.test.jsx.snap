// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShopHeader component should render a category heading and description 1`] = `
.c0 {
  border-bottom: 3px solid var(--color-grey-9);
  padding-bottom: 20px;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c1 {
  font-family: 'Uni Sans',sans-serif;
  font-size: 1.9rem;
  margin-right: 35px;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-transform: rotate(90deg);
  -ms-transform: rotate(90deg);
  transform: rotate(90deg);
  -webkit-transition-delay: 500ms;
  transition-delay: 500ms;
  -webkit-transition-property: -webkit-transform;
  -webkit-transition-property: transform;
  transition-property: transform;
  background: none;
  cursor: pointer;
  font-size: 0;
}

.c6 {
  margin-top: 15px;
}

.c7 {
  display: inline-block;
  font-size: 1.44rem;
  font-weight: 600;
}

.c8 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  font-weight: 700;
  text-transform: uppercase;
  background: none;
  cursor: pointer;
  font-size: 0;
}

.c8 svg {
  height: auto;
  width: 15px;
  -webkit-transform: rotate(90deg);
  -ms-transform: rotate(90deg);
  transform: rotate(90deg);
  -webkit-transition-delay: 500ms;
  transition-delay: 500ms;
  -webkit-transition-property: -webkit-transform;
  -webkit-transition-property: transform;
  transition-property: transform;
}

.c4 {
  color: var(--color-grey-8);
  font-size: 0.88rem;
  line-height: 1.7;
  padding-top: 15px;
}

@media screen and (min-width:0px) {
  .c3 {
    display: none;
  }
}

@media screen and (min-width:650px) {
  .c3 {
    display: block;
  }
}

@media screen and (min-width:0px) {
  .c5 {
    display: block;
  }
}

@media screen and (min-width:650px) {
  .c5 {
    display: none;
  }
}

@media screen and (min-width:650px) {
  .c2 {
    display: none;
  }
}

@media screen and (max-width:649px) {
  .c6 {
    -webkit-align-items: center;
    -webkit-box-align: center;
    -ms-flex-align: center;
    align-items: center;
    display: -webkit-box;
    display: -webkit-flex;
    display: -ms-flexbox;
    display: flex;
    -webkit-box-pack: justify;
    -webkit-justify-content: space-between;
    -ms-flex-pack: justify;
    justify-content: space-between;
  }
}

@media screen and (min-width:700px) {
  .c7 {
    margin-right: 20px;
  }
}

@media screen and (min-width:650px) {
  .c8 {
    -webkit-column-gap: 7px;
    column-gap: 7px;
    color: var(--color-blue-1);
    font-size: 0.8rem;
  }

  .c8 svg {
    height: auto;
    width: 10px;
  }
}

<section>
  <div
    className="c0"
  >
    <h1
      className="c1"
    />
    <button
      aria-controls="top-text"
      aria-expanded={false}
      className="c2"
      data-testid="description-button"
      onClick={[Function]}
      title="Read More"
      type="button"
    >
      Read More
      <svg
        fill="currentColor"
        height={27}
        viewBox="0 0 384 512"
        width={20}
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M342.6 233.4c12.5 12.5 12.5 32.8 0 45.3l-192 192c-12.5 12.5-32.8 12.5-45.3 0s-12.5-32.8 0-45.3L274.7 256 105.4 86.6c-12.5-12.5-12.5-32.8 0-45.3s32.8-12.5 45.3 0l192 192z"
        />
      </svg>
    </button>
  </div>
  <div
    className="c3"
  >
    <p
      className="c4"
    />
  </div>
  <div
    className="c5"
  />
  <div
    className="c6"
  >
    <h2
      className="c1 c7"
    >
      Latest Arrivals - 
    </h2>
    <button
      aria-controls="bottom-text"
      aria-expanded={false}
      className="c8"
      data-testid="latest-arrivals-button"
      onClick={[Function]}
      title="Read More"
      type="button"
    >
      Read More
      <svg
        fill="currentColor"
        viewBox="0 0 384 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M342.6 233.4c12.5 12.5 12.5 32.8 0 45.3l-192 192c-12.5 12.5-32.8 12.5-45.3 0s-12.5-32.8 0-45.3L274.7 256 105.4 86.6c-12.5-12.5-12.5-32.8 0-45.3s32.8-12.5 45.3 0l192 192z"
        />
      </svg>
    </button>
  </div>
</section>
`;
